lib/bolt/module_installer.rb

```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+require 'bolt/puppetfile'
+require 'bolt/puppetfile/installer'
+require 'bolt/error'
+
+module Bolt
+  class ModuleInstaller
+    def initialize(outputter, pal)
+      @outputter = outputter
+      @pal       = pal
+    end
+
+    # Adds a single module to the project.
+    #
+    def add(name, modules, path, moduledir, config)
+      # If the project configuration file already includes this module,
+      # exit early.
+      puppetfile  = Bolt::Puppetfile.new(modules)
+      new_module  = Bolt::Puppetfile::Module.from_hash('name' => name)
+
+      if puppetfile.modules.include?(new_module)
+        @outputter.print_message "Project configuration file #{config} already includes "\
+                                 "module #{new_module}. Nothing to do."
+        return 0
+      end
+
+      # If the Puppetfile exists, make sure it's managed by Bolt.
+      if path.exist?
+        assert_managed_puppetfile(puppetfile, path)
+      end
+
+      # Create a Puppetfile object that includes the new module and its
+      # dependencies. We error early here so we don't add the new module to the
+      # project config or modify the Puppetfile.
+      puppetfile = add_new_module_to_puppetfile(new_module, modules, path)
+
+      # Add the module to the project configuration.
+      @outputter.print_message "Updating project configuration file at #{config}"
+
+      data = Bolt::Util.read_yaml_hash(config, 'project')
+      data['modules'] ||= []
+      data['modules'] <<  { 'name' => new_module.title }
+
+      begin
+        File.write(config, data.to_yaml)
+      rescue SystemCallError => e
+        raise Bolt::FileError(
+          "Unable to update project configuration file: #{e.message}",
+          config
+        )
+      end
+
+      # Write the Puppetfile.
+      @outputter.print_message "Writing Puppetfile at #{path}"
+      puppetfile.write(path, force: true)
+
+      # Install the modules.
+      install_puppetfile(path, moduledir)
+    end
+
+    # Creates a new Puppetfile that includes the new module and its dependencies.
+    #
+    private def add_new_module_to_puppetfile(new_module, modules, path)
+      @outputter.print_message "Resolving module dependencies, this may take a moment"
+
+      # If there is an existing Puppetfile, add the new module and attempt
+      # to resolve. This will not update the versions of any installed modules.
+      if path.exist?
+        puppetfile = Bolt::Puppetfile.parse(path)
+        puppetfile.add_modules(new_module)
+
+        begin
+          puppetfile.resolve
+          return puppetfile
+        rescue Bolt::Error
```

Comment:
Ah, good idea. Added a debug statement in the rescue. To me it makes a bit more sense there than after, since if the second resolve fails it will issue an error detailing which modules have incompatible version requirements. I think that error makes less sense without the debug message saying Bolt will attempt to re-resolve from the config.